crates/ra_hir/src/adt.rs

```diff
@@ -36,19 +38,51 @@ impl Struct {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DeriveData {}
+
+impl DeriveData {
+    pub fn new(_tt: TokenTree) -> Self {
+        DeriveData {}
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AttrData {
+    Derive(DeriveData),
+    MustUse(Option<SmolStr>),
+    Unknown,
+}
+
+impl AttrData {
+    pub fn new(attr: AttrKind) -> Self {
+        match attr {
+            AttrKind::Derive(tt) => AttrData::Derive(DeriveData::new(tt)),
+            AttrKind::MustUse(s) => AttrData::MustUse(s),
+            _ => AttrData::Unknown,
+        }
+    }
+}
+
 /// Note that we use `StructData` for unions as well!
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct StructData {
     pub(crate) name: Option<Name>,
     pub(crate) variant_data: Arc<VariantData>,
+    pub(crate) attr_data: Arc<Vec<AttrData>>,
```

Comment:
Hm, I think instead of adding this as a field to all these structs directly, I think it makes more sense to handle this similarly to generics: Have a [separate query](https://github.com/rust-analyzer/rust-analyzer/blob/0ca30b6c4bb717e3cc54bfc98ffaa7a6d8c2cdfd/crates/ra_hir/src/generics.rs#L61) `attr_data(ModuleDef) -> Arc<Vec<AttrData>>` (maybe wrap the vec in some struct). The code model structs like `Struct` then just implement a trait `HasAttributes` that uses this query to return the attribute data.

Reply:
Hm, thinking about this more, seems like we should process attributes earlier. We only get `ModuleDef`s *after* macro expansion, but attributes **are** macros and can influence the expansion.  In particular, we want to process derives at the same time when we build `CrateDefMap`.

So, it seems like we, unfortunately, will need to slightly different code-paths: an early path to expand attribute-like macros, and a late one to process things like derives.

I suggest focusing on the derive case first, as it seems harder and more fundamental. Derives should be handled mostly like macros:

* first, we should record the fact that derive exists into `RawItems`. I *think* it makes sense to store them with other macros
* then, we should enhance `MacroCallId` to allow it to refer not only to macro_rules invocations, but to attribute macros as well
* then, in `macro_expand_query`,  we should handle this new flavor of macros